File: dotnet/src/Microsoft.Agents.Protocol/AgentProtocolClient.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;

namespace Microsoft.Agents.Protocol;

/// <summary>
/// Main client for interacting with Agent Protocol APIs
/// Provides access to Agents, Runs, and Threads operations
/// </summary>
public class AgentProtocolClient : IDisposable
{
    private readonly HttpClient _httpClient;
    private readonly bool _disposeHttpClient;
    private readonly AgentProtocolClientOptions _options;

    /// <summary>
    /// Gets the Runs API client
    /// </summary>
    public RunsClient Runs { get; }

    /// <summary>
    /// Gets the Threads API client
    /// </summary>
    public ThreadsClient Threads { get; }

    /// <summary>
    /// Gets the Agents API client
    /// </summary>
    public AgentsClient Agents { get; }

    /// <summary>
    /// Creates a new instance of the Agent Protocol client
    /// </summary>
    /// <param name="options">Client configuration options</param>
    public AgentProtocolClient(AgentProtocolClientOptions options)
    {
        _options = options ?? throw new ArgumentNullException(nameof(options));

        if (_options.HttpClient != null)
        {
            _httpClient = _options.HttpClient;
            _disposeHttpClient = false;
        }
        else
        {
            _httpClient = new HttpClient
            {
                BaseAddress = _options.BaseUrl,
                Timeout = TimeSpan.FromSeconds(_options.TimeoutSeconds)
            };
            _disposeHttpClient = true;
        }

        // Configure authentication if API key is provided
        if (!string.IsNullOrEmpty(_options.ApiKey))
        {
            _httpClient.DefaultRequestHeaders.Authorization =
                new AuthenticationHeaderValue("Bearer", _options.ApiKey);
        }

        // Set default headers
        _httpClient.DefaultRequestHeaders.Accept.Add(
            new MediaTypeWithQualityHeaderValue("application/json"));

        // Initialize API clients
        Runs = new RunsClient(_httpClient, _options);
        Threads = new ThreadsClient(_httpClient, _options);
        Agents = new AgentsClient(_httpClient, _options);
    }

    /// <summary>
    /// Creates a new instance with just a base URL
    /// </summary>
    /// <param name="baseUrl">Base URL for the Agent Protocol API</param>
    public AgentProtocolClient(string baseUrl)
        : this(new AgentProtocolClientOptions
        {
            BaseUrl = new Uri(baseUrl)
        })
    {
    }

    /// <summary>
    /// Creates a new instance with base URL and API key
    /// </summary>
    /// <param name="baseUrl">Base URL for the Agent Protocol API</param>
    /// <param name="apiKey">API key for authentication</param>
    public AgentProtocolClient(string baseUrl, string apiKey)
        : this(new AgentProtocolClientOptions
        {
            BaseUrl = new Uri(baseUrl),
            ApiKey = apiKey
        })
    {
    }

    /// <summary>
    /// Disposes the client and underlying HTTP client if owned
    /// </summary>
    public void Dispose()
    {
        if (_disposeHttpClient)
        {
            _httpClient?.Dispose();
        }
    }
}

File: python/microsoft-agents-protocol-client/microsoft/agents/protocol/client/agent_protocol_client.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

"""Main Agent Protocol client"""

from typing import Optional
import aiohttp

from .client_options import AgentProtocolClientOptions
from .runs_client import RunsClient
from .threads_client import ThreadsClient
from .agents_client import AgentsClient


class AgentProtocolClient:
    """
    Main client for interacting with Agent Protocol APIs
    Provides access to Agents, Runs, and Threads operations
    """

    def __init__(self, options_or_url: AgentProtocolClientOptions | str):
        """
        Creates a new instance of the Agent Protocol client

        Args:
            options_or_url: Client configuration options or base URL string
        """
        # Support passing a URL string directly for convenience (as shown in quickstart)
        if isinstance(options_or_url, str):
            self._options = AgentProtocolClientOptions(base_url=options_or_url)
        else:
            self._options = options_or_url

        self._own_session = self._options.session is None

        if self._options.session:
            self._session = self._options.session
        else:
            timeout = aiohttp.ClientTimeout(total=self._options.timeout_seconds)
            headers = {"Accept": "application/json"}

            if self._options.api_key:
                headers["Authorization"] = f"Bearer {self._options.api_key}"

            self._session = aiohttp.ClientSession(
                base_url=self._options.base_url, timeout=timeout, headers=headers
            )

        # Initialize API clients
        self.runs = RunsClient(self._session, self._options)
        self.threads = ThreadsClient(self._session, self._options)
        self.agents = AgentsClient(self._session, self._options)

    @classmethod
    def from_url(cls, base_url: str, api_key: Optional[str] = None):
        """
        Creates a new instance with just a base URL

        Args:
            base_url: Base URL for the Agent Protocol API
            api_key: Optional API key for authentication
        """
        return cls(AgentProtocolClientOptions(base_url=base_url, api_key=api_key))

    async def close(self):
        """Closes the client and releases resources"""
        if self._own_session and self._session:
            await self._session.close()

    async def __aenter__(self):
        """Context manager entry"""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        await self.close()
